### ISSUE TITLE
`pygmt.Figure.plot()`: `incols` parameter works incorrect for rotated rectangle

### ISSUE THREAD

**Issue by yvonnefroehlich** [GenericMappingTools/pygmt#1892]
**Description of the problem**

The `incols` (`-i`) parameter of the method `pygmt.Figure.plot()` appears to work partly incorrect when using the multi-parameter symbol rotated rectangle (`'j'`). The values to define and plot this symbol are passed through an input file (*.txt) to the `data` parameter.

For a first understanding of the behavior, I did a rough systematic analysis for each of the six (symbol) parameters `x`, `y`, `color`, 
'direction', 'width', and 'height' (see the six figures below). At the moment only for one parameter the passed value is changed at a time (highlighted in red font). The titles of the individual subplots give the order of the input columns and the corresponding values (can be interpretated as the expection).

The figures below show, that partly:
- the parameter is incorrectly plotted: for width and height the panels (a), (b), (c).
- another parameter is incorrectly or differently plotted despite no other value was passed to this parameter: for x, y, color, direction the panels (e) and (f).

File with all six figures (generated outside of _PyGMT_ or python): [rotrect_merge_20220421.pdf](https://github.com/GenericMappingTools/pygmt/files/8533219/rotrect_merge_20220421.pdf)

![rotrect_x](https://user-images.githubusercontent.com/94163266/164515734-a97c09cf-a8d9-46f7-bd6f-ed29856b9352.png)

![rotrect_y](https://user-images.githubusercontent.com/94163266/164515771-b7e8be01-d841-4c55-8aee-d8d9ee2f9409.png)

![rotrect_color](https://user-images.githubusercontent.com/94163266/164515794-286e5ce3-0a0b-4758-ad09-36c76027eb09.png)

![rotrect_direction](https://user-images.githubusercontent.com/94163266/164515830-f2a1e90b-bbdb-430a-8cf2-d0b447bfc5e0.png)

![rotrect_width](https://user-images.githubusercontent.com/94163266/164515850-c1a491be-2150-465c-b57b-4b11d75d05be.png)

![rotrect_height](https://user-images.githubusercontent.com/94163266/164515925-ac327a5c-dfc4-4bda-a127-f5d9761d1027.png)

**Full code that generated the error**

File with test data (please place it in the same folder as the python script): [test_rotrect.txt](https://github.com/GenericMappingTools/pygmt/files/8534399/test_rotrect.txt)

```python
import pygmt as gmt


#-----------------------------------------------------------------------------
#-----------------------------------------------------------------------------
# set general stuff
my_length = 10
my_R = [0, my_length, 0, my_length] # region
my_J = 'X' + str(my_length) + 'c/' + str(my_length) + 'c' # projection
my_B = ['WSne','a1g1','x+lx','y+ly'] # frame of figure
my_W = '3p,black' # pen -> outline
my_S = 'j' # symbole -> rotated rectangle: dir_cchor/width/height
my_fontcol = 'red'
my_fontsize = 15 # in point
my_pngdpi = 360 # resolution of output png

#-----------------------------------------------------------------------------
# make colormap
# Scientic Colour mmaps 7.0.1, F. Crameri 2021
cmap_rotrect = 'roma_rotrect.cpt'
gmt.makecpt(cmap = 'roma',
            output = cmap_rotrect,
            series = '0/10/1',)

#-----------------------------------------------------------------------------
# test data
# contains only one record: 4 5 2 0 3 1
# x_position y_position color_fill direction_cchor width height
data_rect = 'test_rotrect.txt'

parameter_all = 'x','y','color','direction','width','height'
index_all = 0,1,2,3,4,5 # for columns, zero-based
value_all = 4,5,2,0,3,1


#-----------------------------------------------------------------------------
#-----------------------------------------------------------------------------
#%% single plots

for parameter in parameter_all:
    
    for index in index_all:
        
        value = value_all[index]
        my_fname = 'rotrect_' + parameter + str(index)
        
        # make strings for input columns orders and individual subplot titles
        # font color change @;color;text@;; (GMT 6.3)
        if parameter=='x':
            incols_rotrect = str(index) + ',1,2,3,4,5',
            title_rotrect = '+t"incols = \'@;"' + my_fontcol + '";"' + \
                str(index) + '"@;;,1,2,3,4,5\' | values = @;"' + my_fontcol + \
                '";"' + str(value) + '"@;; 5 2 0 3 1 "'
        elif parameter=='y':
            incols_rotrect = '0,' + str(index) + ',2,3,4,5'
            title_rotrect = '+t"incols = \'0,@;"' + my_fontcol + '";"' + \
                str(index) + '"@;;,2,3,4,5\' | values = 4 @;"' + my_fontcol + \
                '";"' + str(value) + '"@;; 2 0 3 1 "'    
        elif parameter=='color':
            incols_rotrect = '0,1,' + str(index) + ',3,4,5',
            title_rotrect = '+t"incols = \'0,1,@;"' + my_fontcol + '";"' + \
                str(index) + '"@;;,3,4,5\' | values = 4 5 @;"' + my_fontcol + \
                '";"' + str(value) + '"@;; 0 3 1 "'    
        elif parameter=='direction':
            incols_rotrect = '0,1,2,' + str(index) + ',4,5',
            title_rotrect = '+t"incols = \'0,1,2,@;"' + my_fontcol + '";"' + \
                str(index) + '"@;;,4,5\' | values = 4 5 2 @;"' + my_fontcol + \
                '";"' + str(value) + '"@;; 3 1 "'
        elif parameter=='width':
            incols_rotrect = '0,1,2,3,' + str(index) + ',5'
            title_rotrect = '+t"incols = \'0,1,2,3,@;"' + my_fontcol + '";"' + \
                str(index) + '"@;;,5\' | values = 4 5 2 0 @;"' + my_fontcol + \
                '";"' + str(value) +  '"@;; 1"'
        elif parameter=='height':
            incols_rotrect = '0,1,2,3,4,' + str(index),
            title_rotrect = '+t"incols = \'0,1,2,3,4,@;"' + my_fontcol + '";"' + \
                str(index) + '"@;;\' | values = 4 5 2 0 3 @;"' + my_fontcol + \
                '";"' + str(value) + '@;;'


        fig = gmt.Figure()    

        gmt.config(FONT_TITLE = str(my_fontsize) + 'p',
                   FONT_ANNOT_PRIMARY = str(my_fontsize+2) + 'p',
                   FONT_LABEL = str(my_fontsize+2) + 'p',)
        
        fig.basemap(
            region = my_R,
            projection = my_J,
            frame = my_B,
            )    
        fig.basemap(
            frame = title_rotrect # plot individual subplot title
            )
        
        fig.plot(
            data = data_rect,
            style = my_S,
            cmap = cmap_rotrect,
            pen = my_W,
            incols = incols_rotrect,
            )   
        
        if parameter=='color' and index==0: # show colorbar only once
            fig.colorbar(
                cmap = cmap_rotrect,
                frame = 'x1',
                position = 'JRB+jRB+o0.5c/0.5c+w7.0c/0.5c+ma',
                box = '+gwhite@30+p1.0p,black+r',
                )

        # fig.show()
        fig.savefig(fname = my_fname + '.eps')
        # fig.savefig(fname = my_fname + '.png', dpi = my_pngdpi)


#-----------------------------------------------------------------------------
#-----------------------------------------------------------------------------
#%% subplots

for parameter in parameter_all:
    
    fig_merge = gmt.Figure()
    
    with fig_merge.subplot(nrows = 2, ncols = 3, subsize = (8,8),
                           frame = 'lbtr', autolabel = '(a)', 
                           title = parameter): # overall title
        
        for pcol in range(6):
            
            with fig_merge.set_panel(panel = pcol):
                
                fig_merge.image(
                    region = [0, 1, 0, 1],
                    imagefile = 'rotrect_' + parameter + str(pcol) + '.eps',
                    position = 'JMC+jMC+w7c',
                    )
                
    fig_merge.show()    
    # fig_merge.savefig(fname = 'rotrect_' + parameter + '.png', dpi = my_pngdpi)
    # fig_merge.savefig(fname = 'rotrect_' + parameter + '.pdf', dpi = my_pngdpi)
```

**Full error message**
No error message occurs.

**System information**

Please paste the output of `python -c "import pygmt; pygmt.show_versions()"`:

_Windows_
```
PyGMT information:
  version: v0.6.1
System information:
  python: 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:17:03) [MSC v.1929 64 bit (AMD64)]
  executable: C:\ProgramData\Anaconda3\envs\pygmt_env\python.exe
  machine: Windows-10-10.0.19043-SP0
Dependency information:
  numpy: 1.22.3
  pandas: 1.4.2
  xarray: 2022.3.0
  netCDF4: 1.5.8
  packaging: 21.3
  geopandas: 0.10.2
  ghostscript: 9.54.0
  gmt: 6.3.0
GMT library information:
  binary dir: C:/ProgramData/Anaconda3/envs/pygmt_env
  cores: 4
  grid layout: rows
  library path: C:/ProgramData/Anaconda3/envs/pygmt_env/Library/bin/gmt.dll
  padding: 2
  plugin dir: C:/ProgramData/Anaconda3/envs/pygmt_env/Library/bin/gmt_plugins
  share dir: C:/Program Files (x86)/gmt6/share
  version: 6.3.0
```

_Linux_
```
PyGMT information:
  version: v0.6.1
System information:
  python: 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:25:59)  [GCC 10.3.0]
  executable: /home/yfroe/.conda/envs/pygmt_env_new/bin/python
  machine: Linux-5.3.18-150300.59.60-default-x86_64-with-glibc2.31
Dependency information:
  numpy: 1.22.3
  pandas: 1.4.2
  xarray: 2022.3.0
  netCDF4: 1.5.8
  packaging: 21.3
  geopandas: 0.10.2
  ghostscript: 9.54.0
  gmt: 6.3.0
GMT library information:
  binary dir: /home/yfroe/.conda/envs/pygmt_env_new/bin
  cores: 20
  grid layout: rows
  library path: /home/yfroe/.conda/envs/pygmt_env_new/lib/libgmt.so
  padding: 2
  plugin dir: /home/yfroe/.conda/envs/pygmt_env_new/lib/gmt/plugins
  share dir: /home/yfroe/.conda/envs/pygmt_env_new/share/gmt
  version: 6.3.0
 ```

**Comment by seisman** [comment]
@yvonnefroehlich Thanks for your detailed issue report! I confirm that I can reproduce the issue.

Here is a minimal example to reproduce the issue using GMT CLI:
```
echo 4 5 2 0 3 1 | gmt plot -R0/10/0/10 -JX10c -Sj -Croma_rotrect.cpt -Ba1f1g1 -i4,1,2,3,4,5 -pdf map
```
The CPT [roma_rotrect.cpt.txt](https://github.com/GenericMappingTools/pygmt/files/8542067/roma_rotrect.cpt.txt) is also attached (need to remove the ".txt" suffix).

@PaulWessel Please see if you can find the fix.

**Comment by yvonnefroehlich** [comment]
@seisman Thanks for your fast answer!

**Comment by PaulWessel** [comment]
Note that @yvonnefroehlich is selecting column 4 twice: First as the x-coordinate and second as the major axis.  But: The major axis is a dimensional entity (cm, inch) that is internally converted to inches while x is a coordinate subject to -J.  Because it is first selected as coordinate it is not converted the inches and remain as 3 - it is then (after conversion) just duplicated to the major axis and no conversion to inches is done. We can declare this a bug but it is not clear how to fix it since it mixes and matches coordinates and lengths.  I will have to think about this.

**Comment by seisman** [comment]
@yvonnefroehlich 

Paul has opened the PR https://github.com/GenericMappingTools/gmt/pull/6616 to fix the issue. Attached here ([rotrect_new.pdf](https://github.com/GenericMappingTools/pygmt/files/8547563/rotrect_new.pdf)) is the new PDF file after the fix. Please let me know if it works what you expect.

**Comment by yvonnefroehlich** [comment]
Thanks @PaulWessel for your fast answer and effort! Due to your explanation I understand, why a color-coding of the bars based on the orientation works, but based on the width fails, when selecting the corresponding columns twice.

Thanks @seisman for generating this new PDF file! Now the bars look in all cases as I expect.

**Comment by seisman** [comment]
@yvonnefroehlich 

The upstream fix has been merged into the GMT's master branch. If you want to try it, you can install the latest **dev** version of GMT using the following command:
```
conda install -c conda-forge/label/dev gmt
```
